### PR TITLE
Disable dictionary processing for non-deterministic expressions

### PR DESCRIPTION
**Pull request by nileema**
If a projection is non-deterministic fall back to simple columnar
processing instead of dictionary processing.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        public Boolean visitConstant(ConstantExpression literal, Void context)
+        {
+            return true;
+        }
+
+        @Override
+        public Boolean visitCall(CallExpression call, Void context)
+        {
+            // TODO: total hack to figure out if a function is deterministic. martint should fix this when he refactors the planning code
+            if (call.getSignature().getName().equals("random") || call.getSignature().getName().equals("rand")) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java]: @erichwang or @martint is there is better way to know if a function is deterministic?  Can we look up in the function registry?

**Comment by dain** [comment]
Can you add a test that verifies that you don't get the a dictionary when the function is non-deterministic?  IIRC we have a test verifies we have get the same dictionary for a projection, this would be the opposite of that test.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java [hunk 2]
```diff
@@ -415,7 +415,7 @@ private static MethodDefinition generateProjectDictionaryMethod(
 
         List<Integer> inputChannels = getInputChannels(projection);
 
-        if (inputChannels.size() != 1) {
+        if (inputChannels.size() != 1 || !projection.isDeterministic()) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageProcessorCompiler.java]: does this disable dictionary for projects and filters?

File: presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java (hunk 1 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java]: This should be done the same way as `ExpressionOptimizer`.

File: presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                return false;
+            }
+
+            boolean result = true;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java]:
```java
return call.getArguments().stream()
        .allMatch(expression -> expression.accept(this, context));

File: presto-main/src/main/java/com/facebook/presto/sql/relational/CallExpression.java [hunk 4]
```diff
@@ -52,6 +52,12 @@ public Type getType()
         return returnType;
     }
 
+    @Override
+    public boolean isDeterministic()
+    {
+        return DeterminismEvaluator.isDeterministic(this);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/CallExpression.java]: This is weird to add an abstract method that always calls a static function. Why not just use the static function directly?

File: presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        @Override
+        public Boolean visitConstant(ConstantExpression literal, Void context)
+        {
+            return true;
+        }
+
+        @Override
+        public Boolean visitCall(CallExpression call, Void context)
+        {
+            if (!registry.getScalarFunctionImplementation(call.getSignature()).isDeterministic()) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/DeterminismEvaluator.java]: IIRC, `CallExpressions` supports signatures that are not "functions" in the `FunctionRegistry`:

https://github.com/facebook/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/gen/BytecodeExpressionVisitor.java#L70

I'm not sure how you know that it should be in the registry.  @martint any ideas?

**Comment by dain** [comment]
Travis is failing with:
```
Caused by: com.facebook.presto.spi.PrestoException: COALESCE(bigint,bigint):bigint not found
	at com.facebook.presto.metadata.FunctionRegistry.getScalarFunctionImplementation(FunctionRegistry.java:628)
	at com.facebook.presto.sql.relational.DeterminismEvaluator$Visitor.visitCall(DeterminismEvaluator.java:59)
```

**Comment by nileema** [comment]
@dain updated. 

@martint suggested that if a function is not registered we should treat it as deterministic.

**Comment by nileema** [comment]
Also added ``testNonDeterministicProject`` to ``TestPageProcessorCompiler``

**Comment by dain** [comment]
Looks good

**Pull request merged**